beluga/include/beluga/sensor/data/regular_grid.hpp

```diff
@@ -67,15 +67,25 @@ class BaseRegularGrid2 : public ciabatta::ciabatta_top<Derived> {
  public:
   /// Compute nearest grid cell coordinates given plane coordinates.
   /**
-   * Note this is a surjective function.
+   * For performance reasons this function is optimized for coordiantes in the
+   * positive quadrant (i.e. valid grid coordinates). For coordinates in other
+   * quadrants a representative cell in the same quadrant will be returned instead.
    *
    * \param x Plane x-axis coordinate.
    * \param y Plane y-axis coordinate.
    * \return Grid cell coordinates.
    */
   [[nodiscard]] Eigen::Vector2i cell_near(double x, double y) const {
-    const auto xi = static_cast<int>(std::floor(x / this->self().resolution()));
-    const auto yi = static_cast<int>(std::floor(y / this->self().resolution()));
+    const auto inv_resolution = 1 / this->self().resolution();
+    const auto scaled_x = x * inv_resolution;
+    const auto scaled_y = y * inv_resolution;
+    // Truncate towards zero. Use a poor man's replacement of
+    // std::floor for coodinates in the positive quadrant because
+    // the `std::floor` is awfully slow and this function is
+    // performance critical. We don't care about coordinates in other
+    // quadrants so we just return some equally invalid representative.
+    const auto xi = scaled_x < 0.0 ? -1 : static_cast<int>(scaled_x);
+    const auto yi = scaled_y < 0.0 ? -1 : static_cast<int>(scaled_y);
```

Comment:
@glpuga wouldn't:
```suggestion
    const auto xi = static_cast<int>(scaled_x) - (scaled_x < 0.0);
    const auto yi = static_cast<int>(scaled_y) - (scaled_y < 0.0);
```
be a correct `floor` implementation at a negligible extra cost (that of two integer substractions)?

Reply:
I can't think of a case where it would not work. At least it does not have the problem of https://stackoverflow.com/questions/824118/why-is-floor-so-slow/825916#825916

Code looks good too; it's branch-less and the compiler exploits that to vectorize lines 87 and 88.
https://godbolt.org/z/Gzssvcr76

Reply:
I'm wrong. It does provide the wrong answer right in the threshold of negative integers. It's the same problem as the first link.

But as far as wrong answers go for out-of-grid coordinates, I like the compiled code of this version better.